Randomize the order of peers in peerlists.

echohead opened the pull request:
Previously, peers would always be used in the order as input.

This can lead to non-uniform load on the peers in some cases.
For example:
- rpc callers are crash looping and repeatedly coming back up
  and hitting the first peer. (first peer will be hammered)
- many callers are started around the same time, and perform
  rpcs on a fixed schedule (each peer will be hammered in turn)

Instead, randomly shuffle peers as they are added.

By default, the random-number generator used for shuffling is
seeded with the current time in nanoseconds.

Support an Option for setting the random seed, to allow for
deterministic tests.

- [x] Description and context for reviewers: one partner, one stranger
- [ ] Docs (package doc)
- [x] Entry in CHANGELOG.md

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/yarpc/yarpc-go?pullRequest=1422) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/yarpc/yarpc-go?pullRequest=1422) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/yarpc/yarpc-go?pullRequest=1422) it.</sub>

kriskowal commented on peer/pendingheap/list_test.go:
```diff
@@ -149,13 +149,13 @@ func TestPeerHeapList(t *testing.T) {
 				UpdateAction{AddedPeerIDs: []string{"1", "2", "3", "4", "5", "6"}},
-				ChooseAction{ExpectedPeer: "1"},
-				ChooseAction{ExpectedPeer: "2"},
+				ChooseAction{ExpectedPeer: "6"},
```
Would it be possible to disable shuffling for this test? The shuffle eliminates some clarity that this test verifies that the fewest-pending-requests peer selector degenerates to round-robin when load is perfectly even.

kriskowal commented on peer/peerlist/list.go:
```diff
@@ -568,3 +584,24 @@ func (pl *List) Introspect() introspection.ChooserStatus {
 	}
 }
+
+// shuffle randomizes the order of a slice of peers.
```
May as well comment that this is Fisher-Yates.

echohead replied:
done.

kriskowal commented:
Oh, and please leave an entry in the change log.

echohead commented:
- added an option to disable shuffle
- updated pendingheap test to use `Shuffle(false)` and reverted changes changes to pendingheap tests to preserve clarity.
- updated changelog.

echohead commented:
- don't make any additions to public API of round-robin and pendingheap -- just add private Options to control shuffling for use by tests.
- have roundrobin tests disable shuffle to preserve clear ordering.
- add a single roundrobin test which leaves on the default shuffle behavior for code coverage.

willhug commented on peer/peerlist/list.go:
```diff
@@ -66,6 +72,24 @@ func Capacity(capacity int) ListOption {
+// Shuffle enables/disables shuffling the order of peers as they are added.
+//
+// Default is true.
+func Shuffle(shuffle bool) ListOption {
```
If the default is true, we just just make this a `DisableShuffling()` option instead of taking a boolean.  Makes it easier to reason about what is happening.

echohead replied:
changed option signature to be `func NoShuffle() ListOption`, with no argument.

echohead commented on peer/peerlist/list.go:
```diff
@@ -66,6 +72,22 @@ func Capacity(capacity int) ListOption {
 }
 
+// NoShuffle disables the default behavior of shuffling peerlist order.
+func NoShuffle() ListOption {
```
done.

echohead commented on peer/pendingheap/list_test.go:
```diff
@@ -611,3 +612,9 @@ func TestPeerHeapList(t *testing.T) {
 	}
 }
+
+func shuffle(shuffle bool) ListOption {
```
done.

echohead commented on peer/roundrobin/list_test.go:
```diff
@@ -970,3 +996,15 @@ func (p *testPeer) Status() peer.Status {
 
 func (*testPeer) EndRequest() {}
+
+func shuffle(shuffle bool) ListOption {
```
done.

prashantv commented on peer/peerlist/list.go:
```diff
@@ -66,6 +72,20 @@ func Capacity(capacity int) ListOption {
 }
 
+// NoShuffle disables the default behavior of shuffling peerlist order.
+var NoShuffle listOptionFunc = func(options *listOptions) {
```
one minor issue with using `NoShuffle` instead of `NoShuffle()` is that it's a `var` that anyone can replace. For consistency with other options, and for better docs, I think it's better to keep it a function.

E.g., `NoShuffle` doesn't show up under options like the others:
https://godoc.org/go.uber.org/yarpc/peer/peerlist

kriskowal replied:
Looks like @echohead landed it as-is. :+1:

prashantv replied:
Since it hasn't been released yet, can we update it before release?
